src/proptest.rs

```diff
@@ -0,0 +1,95 @@
+use std::iter::zip;
+
+use crate::{Direction, PortGraph, PortIndex};
+use proptest::prelude::*;
+use rand::seq::SliceRandom;
+
+prop_compose! {
+    /// A random graph with random number of ports but no edges
+    ///
+    /// The graph has at least one node.
+    ///
+    ///  - `max_n_nodes` is the maximum number of nodes in the generated graph
+    ///  - `max_port` is the maximum number of incoming and outgoing ports at
+    ///    every node
+    fn gen_no_edge_graph(max_n_nodes: usize, max_port: usize)(
+        ports in prop::collection::vec(0..=max_port, 2..=max_n_nodes)
+    ) -> PortGraph {
+        let mut g = PortGraph::new();
+        let mut ind = 0;
+        while ind + 1 < ports.len() {
+            g.add_node(ports[ind], ports[ind + 1]);
+            ind += 2;
+        }
+        g
+    }
+}
+
+/// A random graph and random edge lists
+///
+/// Returns a tuple of
+///  - `graph`, a graph with ports but without edges
+///  - `in_ports`, a list of incoming ports
+///  - `out_port`, a list of outgoing ports, of the same length as `in_ports`
+fn gen_graph_and_edges(
+    max_n_nodes: usize,
+    max_port: usize,
+    max_n_edges: usize,
+) -> impl Strategy<Value = (PortGraph, Vec<PortIndex>, Vec<PortIndex>)> {
+    let graph = gen_no_edge_graph(max_n_nodes, max_port);
+    (0..=max_n_edges, graph).prop_perturb(|(mut n_edges, graph), mut rng| {
+        let mut in_ports = Vec::new();
+        let mut out_ports = Vec::new();
+        for p in graph.ports_iter() {
+            match graph.port_direction(p).unwrap() {
+                Direction::Incoming => in_ports.push(p),
+                Direction::Outgoing => out_ports.push(p),
+            }
+        }
```

Comment:
Just nitpicking, but check `Iterator::partition` next time :)

Reply:
Oh excellent, super useful! Thank you for pointing that out, please keep teaching me things :)